source/agora/common/TransactionPool.d

```diff
@@ -204,20 +213,66 @@ public class TransactionPool
 
     ***************************************************************************/
 
-    public int opApply (scope int delegate(Hash hash, Transaction tx) dg) @trusted
+    public int opApply (DG1T) (scope DG1T dg)
+        if (is(DG1T : int delegate(Transaction)) ||
+            is(DG1T : int delegate(ref Transaction)) ||
+            is(DG1T : int delegate(const Transaction)) ||
+            is(DG1T : int delegate(const ref Transaction)))
     {
-        auto results = this.db.execute("SELECT key, val FROM tx_pool");
-
-        foreach (row; results)
+        () @trusted
         {
-            auto hash = *cast(Hash*)row.peek!(ubyte[])(0).ptr;
-            auto tx = deserializeFull!Transaction(row.peek!(ubyte[])(1));
+            auto results = this.db.execute("SELECT val FROM tx_pool");
+
+            foreach (ref row; results)
+            {
+                auto tx = deserializeFull!Transaction(row.peek!(ubyte[])(0));
+
+                // break
+                if (auto ret = dg(tx))
+                    return ret;
+            }
+            return 0;
+        }();
 
-            // break
-            if (auto ret = dg(hash, tx))
-                return ret;
+        // HACK: We don't want this code to ever execute and the dg should never
+        // be `null` (it would have segfaulted in the iteration).
+        // However, we want the delegate attributes to determine the `opApply`'s
+        // attribute, hence this block.
+        // Remove when the SQLite binding is `@safe` and the `@trusted` delegate
+        // can be removed
+        if (dg is null)
+        {
+            Transaction tx;
+            dg(tx);
         }
+        return 0;
+    }
+
+    public int opApply (DG2T) (scope DG2T dg)
```

Comment:
If you annotate this overload with `@disable` you'll see it's never actually used anymore. So we may want to just get rid of it?

Reply:
Yes, it compiles with `@disable`.
Does it mean that the overload for `(Transaction)` works for `(Hash, Transaction)` too?

Reply:
No it means it's not used anymore! Because you changed all of the calls to only use `tx`.

Previously, for whatever reason we used `foreach (hash, tx)` in many places, but then discarded the hash. But you've corrected that, and now they're just doing `foreach (tx) ...`. So you don't need the second overload anymore, just the one taking transactions.